aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -283,11 +287,85 @@ namespace detail {
 template<class... Args> inline void unused_arg_(const Args&...) {}
 }
 
+template <typename T>
+inline c10::IValue get_ivalue_copy(const T& v) {
+  return c10::IValue();
+}
+
+#define IVALUE_COPY_FOR_TYPE(T) \
+    template <> \
+    inline c10::IValue get_ivalue_copy<T>(const T& v) { \
+      return v; \
+    }
+
+IVALUE_COPY_FOR_TYPE(at::Tensor);
+IVALUE_COPY_FOR_TYPE(at::Scalar);
+IVALUE_COPY_FOR_TYPE(double);
+IVALUE_COPY_FOR_TYPE(int64_t);
+IVALUE_COPY_FOR_TYPE(int32_t);
+IVALUE_COPY_FOR_TYPE(bool);
+
+// currently not working with incomplete type
+template <typename T, std::enable_if_t<c10::impl::not_ok_to_box<T>::value>* = nullptr>
+inline bool push_ivalue_copy(std::vector<c10::IValue>& stack, const T& v) {
+  // don't do anything with stack
+  return false;
+}
+template <typename T, std::enable_if_t<!c10::impl::not_ok_to_box<T>::value>* = nullptr>
+inline bool push_ivalue_copy(std::vector<c10::IValue>& stack, const T& v) {
+  torch::jit::push(stack, v);
+  return true;
+}
+
+// attempt to workaround some(?) incomplete types
+
+// complete and possible to instantiate
+template<typename T, typename F = void>
+struct is_instanceable {
+  static const bool value = false;
+};
+
+template<typename T>
+struct is_instanceable<T, std::void_t<decltype(sizeof(T()))>> {
+  static const bool value = true;
+};
+
+template <typename T, std::enable_if_t<!is_instanceable<T>::value>* = nullptr>
+inline bool push_ivalue_copy_2(std::vector<c10::IValue>& stack, const T& v) {
+  // don't do anything with stack
+  return false;
+}
+
+template <typename T, std::enable_if_t<is_instanceable<T>::value>* = nullptr>
+inline bool push_ivalue_copy_2(std::vector<c10::IValue>& stack, const T& v) {
+  bool to_push = (std::is_constructible<IValue, T>::value &&
+      std::is_convertible<T, IValue>::value) || std::is_same<TensorOptions, T>::value;
+  if (to_push) {
+    torch::jit::push(stack, v);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+
 template<class Return, class... Args>
 inline Return Dispatcher::callUnboxedWithDispatchKey(const OperatorHandle& op, DispatchKey dispatchKey, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
   const auto& dispatchTable = op.operatorIterator_->op.dispatch_table();
   const KernelFunction& kernel = dispatch_(dispatchTable, dispatchKey);
+
+  // RECORD_FUNCTION macro will attempt to box the arguments only if we have active observers enabled
+  RECORD_FUNCTION(op.schema().name(), std::vector<c10::IValue>{get_ivalue_copy(args)...}, at::FunctionSequenceNumber::peek());
```

Comment:
Isn't this going to `RECORD_FUNCTION` for every key in the dispatch chain? Is that desired?

Reply:
yes, that's the plan